src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Controls/PopupControlService.cs

```diff
@@ -236,557 +221,557 @@ private void ProcessKeyUp(object sender, KeyEventArgs e)
-            bool foundToolTip = false;
-            showToolTip = false;
+            Debug.Assert(_currentToolTip == null);
+            ResetCurrentToolTipTimer();
+            OnForceClose(null, EventArgs.Empty);
+
+            bool show = true;
 
+            IInputElement element = o as IInputElement;
             if (element != null)
             {
-                FindToolTipEventArgs args = new FindToolTipEventArgs(triggerAction);
+                ToolTipEventArgs args = new ToolTipEventArgs(opening:true);
+                // ** Public callout - re-entrancy is possible **//
                 element.RaiseEvent(args);
 
-                foundToolTip = args.Handled;
+                // [re-examine _currentToolTip, re-entrancy can change it]
+                show = !args.Handled && (_currentToolTip == null);
+            }
 
-                if (args.TargetElement != null)
+            if (show)
+            {
+                object tooltip = ToolTipService.GetToolTip(o);
+                ToolTip tip = tooltip as ToolTip;
+                if (tip != null)
+                {
+                    _currentToolTip = tip;
+                }
+                else
                 {
-                    // Open this element's ToolTip
-                    o = args.TargetElement;
-                    showToolTip =  true;
+                    _currentToolTip = new ToolTip();
+                    _currentToolTip.SetValue(ServiceOwnedProperty, BooleanBoxes.TrueBox);
+
+                    // Bind the content of the tooltip to the ToolTip attached property
+                    Binding binding = new Binding();
+                    binding.Path = new PropertyPath(ToolTipService.ToolTipProperty);
+                    binding.Mode = BindingMode.OneWay;
+                    binding.Source = o;
+                    _currentToolTip.SetBinding(ToolTip.ContentProperty, binding);
                 }
-                else if (args.KeepCurrentActive)
+
+                if (!_currentToolTip.StaysOpen)
                 {
-                    // Keep the current ToolTip active
-                    o = null;
-                    showToolTip =  true;
+                    // The popup takes capture in this case, which causes us to hit test to the wrong window.
+                    // We do not support this scenario. Cleanup and then throw and exception.
+                    throw new NotSupportedException(SR.Get(SRID.ToolTipStaysOpenFalseNotAllowed));
                 }
+
+                _currentToolTip.SetValue(OwnerProperty, o);
+                _currentToolTip.Opened += OnToolTipOpened;
+                _currentToolTip.Closed += OnToolTipClosed;
+                _currentToolTip.FromKeyboard = fromKeyboard;
+                _currentToolTip.IsOpen = true;
+
+                CurrentToolTipTimer = new DispatcherTimer(DispatcherPriority.Normal);
+                CurrentToolTipTimer.Interval = TimeSpan.FromMilliseconds(ToolTipService.GetShowDuration(o));
```

Comment:
We change the  ShowDuration default value from 5000 to int.MaxValue in ToolTipService.cs . And I think we can check the  ToolTipService.GetShowDuration result first, because we will do nothing when it is int.MaxValue

Reply:
This is defensive programming, always starting the timer even when it won't tick for 25 days.  Code doesn't have to check for special cases (including whether CurrentToolTipTimer is null).